Add adaptive scientific notation for frequency range

rahul-flex opened the pull request:
As discussed with @tomflexcompute we want to include as many significant digits as possible, provided the two numbers (frequency range: fmin_med and fmax_med) remain identical up to that point.
Once a differing digit is encountered, the output rounds off subsequent digits. 
Also, the function returns the min and max values with same exponent.


Below are few test examples:
inputs are fmin_med and fmax_med


`Input: 1234567.56, 1234577.34
Output: 1.23457e6, 1.23458e6

Input: 1234567.56, 211234577.34
Output: 1e6, 211e6

Input: 1234567, 1934577
Output: 1e6, 2e6

Input: 1234567, 1284577
Output: 1.2e6, 1.3e6

Input: 123, 128
Output: 1.2e2, 1.3e2

Input: 0.00123, 0.00128
Output: 1.2e-3, 1.3e-3

Input: 0, 10
Output: 0e+00, 1e+01

Input: 123456789012345, 123456789012346
Output: 1.23456789012345e14, 1.23456789012346e14`

rahul-flex committed:
feat: Add scientific notation for frequency range

tomflexcompute commented:
@rahul-flex actually I think we should at least display a certain number of digitals like 5 for example. Since in most materials, the f_min and f_max differ in the first significant digital already. Could you test it on the test model and see how the output warning looks like?

rahul-flex commented:
@tomflexcompute The warning message for the test setup you provided looks like this:
```
18:15:58 EST WARNING: The medium associated with structures[0] has a frequency  
range: (1e12, 250e12) (Hz) that does not fully cover the           
frequencies contained in monitors[0]. This can cause inaccuracies  
in the recorded results. 
```

tomflexcompute commented:
Yeah I think we do need more significant digitals otherwise the numbers are too crude.

Fyi this pr is to address https://github.com/flexcompute/tidy3d/issues/1086

momchil-flex commented on tidy3d/components/simulation.py:
```diff
@@ -2805,6 +2805,34 @@ def warn(istruct, side):
 
         return val
 
+    def _scientific_notation(fmin_med: float, fmax_med: float) -> Tuple[str, str]:
```
I guess this shouldn't be a Simulation level method. Maybe Tidy3dBaseModel? Let's also change the variable names to not specifically refer to frequency, more like `min_val, max_val`? And then regarding @tomflexcompute's point, add an extra argument like `min_digits=4`?  Looking at the code it seems that this number would refer not to the total number of significant digits, but to the digits after the decimal, e.g. `1.1234e12 to 250.1234e12` right? I think that's fine.

yaugenst-flex replied:
agree with `Tidy3dBaseModel` ([here](https://github.com/flexcompute/tidy3d/blob/develop/tidy3d/components/base.py)). also decorate this with a `staticmethod` and change the below calls to `self._scientific_notation` instead of `cls._scientific_notation`, otherwise this will error when called on a class instance.

yaugenst-flex commented on tidy3d/components/simulation.py:
```diff
@@ -2805,6 +2805,34 @@ def warn(istruct, side):
+
+        precision = 0
+        while round(normalized_fmin, precision) == round(normalized_fmax, precision):
+            precision += 1
```
Edge case: this will cause an infinite loop if `normalized_fmin == normalized_fmax`. Not very likely to happen, but should be handled

yaugenst-flex reviewed (commented):
Thanks @rahul-flex this looks pretty good overall. Don't forget to add a unit test and a changelog entry though, this is a nicely testable function.

rahul-flex commented:
Thanks, @momchil-flex, @tomflexcompute, and @yaugenst-flex for the review comments. I have updated the function and added test cases.
@momchil-flex Yes, the min_digits refers to the digits after the decimal in scientific notation.

tomflexcompute reviewed (approved):
Thanks @rahul-flex I think it should work very well.

yaugenst-flex commented on tests/test_components/test_base.py:
Thanks @rahul-flex the test is great, only request I have is could you turn this into a parameterized test instead? That makes them run as separate tests so it will be easier to debug in case something fails. Also note that you generally don't need to add assertion messages in `pytest` because they are a bit redundant as you will get a summary of the failed tests and their parameters.

Here is how to convert to a parameterized test:
```python

@pytest.mark.parametrize(
    "min_val, max_val, min_digits, expected",
    [
        (1234567, 1234577, 4, ("1.23457e6", "1.23458e6")),
        (1234567, 1234577, 6, ("1.234567e6", "1.234577e6")),
        (1.23e-3, 1.28e-3, 4, ("1.2300e-3", "1.2800e-3")),
        (123456789012345, 123456789012346, 4, ("1.23456789012345e14", "1.23456789012346e14")),
        (123656789012345, 123756789012346, 4, ("1.2366e14", "1.2376e14")),
        (123, 123, 4, ("1.2300e2", "1.2300e2")),
    ],
    ids=[
        "default_min_digits",
        "increased_min_digits",
        "small_numbers",
        "large_numbers_precise",
        "large_numbers_rounded",
        "identical_numbers"
    ]
)
def test_scientific_notation(min_val, max_val, min_digits, expected):
    """Test the _scientific_notation method with various inputs."""
    result = Tidy3dBaseModel._scientific_notation(min_val, max_val, min_digits=min_digits)
    assert result == expected
```

yaugenst-flex commented on tidy3d/components/simulation.py:
```diff
@@ -2833,6 +2833,10 @@ def _warn_monitor_mediums_frequency_range(cls, val, values):
 
                     # make sure medium frequency range includes all monitor frequencies
                     fmin_med, fmax_med = medium.frequency_range
+                    sci_fmin_med, sci_fmax_med = Tidy3dBaseModel._scientific_notation(
```
`Simulation` is a subclass of `Tidy3dBaseModel`, so just call `self._scientific_notation()` here

yaugenst-flex reviewed (commented):
Thanks @rahul-flex almost good to go :)

yaugenst-flex reviewed (approved):
Nice, looks great @rahul-flex!

momchil-flex commented on CHANGELOG.md:
```diff
@@ -27,6 +27,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [2.8.0rc1] - 2024-12-17
 
 ### Added
+- Scientific notation for frequency range warnings.
```
This changelog item should be in Unreleased not in 2.8.0rc1. Also generally we tend to put new things at the *bottom* of the list (Added/Fixed/Changed), not at the top.

momchil-flex reviewed (approved):
Looks good after small changelog comment, and eventually squashing all commits.
